astropy/io/votable/tree.py

```diff
@@ -2527,6 +2530,22 @@ def fields(self):
         """
         return self._fields
 
+    @property
+    def all_fields(self):
+        """
+        A list of :class:`Field` objects describing the types of each
+        of the data columns. Contrary to ``fields``, this property should
+        list every field that's available on disk, including deselected columns.
+        """
+        # since we support extending self.fields directly via list.extend
+        # and list.append, self._all_fields may go out of sync.
+        # To remedy this issue, we sync back the public property upon access.
+        for field in self._fields:
+            if field not in self._all_fields:
+                self._all_fields.append(field)
```

Comment:
Sorry, one more thought... Does the order matter, @tomdonaldson ? Just appending might not preserve original order.

Reply:
Yes, order matters.
I'll admit that it's not evident that this bit works properly because it relies on *other* parts of the module to be correct too. The intention is that `all_fields` follows the order of on-disk fields and *then* the insertion order of manually added fields. In principle, `self.fields` can only contain a similarly ordered subset of `self.all_fields`, but a bug in how `self.fields` is constructed would cascade into an incorrect ordering in `self.all_fields` too.

Reply:
I am starting to ponder if we should just bite the bullet and fix this properly as Tom pointed out in https://github.com/astropy/astropy/pull/15959#discussion_r1569177870 ... 🤔

Reply:
Can we agree to explore it in a follow up PR so this patch doesn't get too bloated with deprecation logic ?

Reply:
I am worried that the non-deterministic order would come back and bite us as a even more subtle bug.

Reply:
But if @tomdonaldson can tell me that my worry is unfounded, then I am okay with merging this and have the other idea as follow-up, though if we do that anyway, then this change would be mostly reverted, so is it worth merging and then reverting? 🤪

Reply:
But it *is* deterministic ? To be clear, I'm proposing to get the follow up deprecation patch going as soon as this one gets merged, not in an undetermined future, so there won't be much opportunities for this to break or even to stay brittle.

Reply:
Sorry I didn't see your second message

> though if we do that anyway, then this change would be mostly reverted, so is it worth merging and then reverting? 🤪

It can't be reverted until `table.fields.append/extend` is disallowed, which requires a deprecation period